File: nexus-platform/services/api-gateway/python/src/routers/proxy.py
# =============================================================================
# NEXUS PLATFORM - API GATEWAY - PROXY ROUTER
# =============================================================================
# Dynamic reverse proxy for routing requests to backend services.
# =============================================================================

"""
Proxy Router

This module implements the core proxy functionality of the API Gateway,
routing requests to appropriate backend services based on path mapping.

Features:
- Path-based routing
- Request/response transformation
- Header forwarding
- Error handling
- Retry logic
- Streaming response support
"""

import time
from typing import Dict, Optional

import httpx
import structlog
from fastapi import APIRouter, Request, Response
from fastapi.responses import StreamingResponse

from ..services.service_discovery import ServiceDiscovery
from ..middleware.circuit_breaker import with_circuit_breaker
from ..middleware.metrics import record_upstream_duration
from ..utils.errors import NotFoundError, ServiceUnavailable, TimeoutError

logger = structlog.get_logger(__name__)

router = APIRouter()


# Service routing configuration
# Maps URL path prefixes to backend services
SERVICE_ROUTES: Dict[str, str] = {
    "/auth": "auth-service",
    "/users": "user-service",
    "/payments": "payment-service",
    "/workflows": "workflow-engine",
    "/queue": "distributed-queue",
    "/stream": "streaming-platform",
    "/cache": "cache-layer",
    "/scheduler": "scheduler-service",
    "/workers": "worker-fleet",
}

# Headers to forward from client to backend
FORWARD_HEADERS = {
    "accept",
    "accept-encoding",
    "accept-language",
    "content-type",
    "content-length",
    "user-agent",
    "x-request-id",
    "x-correlation-id",
    "x-forwarded-for",
    "x-forwarded-proto",
    "x-real-ip",
    "authorization",
}

# Headers to exclude from forwarding
EXCLUDED_HEADERS = {
    "host",
    "connection",
    "keep-alive",
    "transfer-encoding",
    "te",
    "trailer",
    "upgrade",
}

# Headers to add to backend requests
def get_proxy_headers(request: Request) -> Dict[str, str]:
    """
    Build headers to forward to backend services.
    
    Args:
        request: The incoming request.
    
    Returns:
        Dict: Headers to forward.
    """
    headers = {}
    
    # Forward allowed headers
    for key, value in request.headers.items():
        key_lower = key.lower()
        if key_lower in FORWARD_HEADERS and key_lower not in EXCLUDED_HEADERS:
            headers[key] = value
    
    # Add gateway-specific headers
    headers["X-Gateway-Request-Id"] = getattr(request.state, "request_id", "unknown")
    
    # Add user info if authenticated
    user = getattr(request.state, "user", None)
    if user:
        headers["X-User-Id"] = user.user_id
        if user.tenant_id:
            headers["X-Tenant-Id"] = user.tenant_id
        if user.roles:
            headers["X-User-Roles"] = ",".join(user.roles)
    
    # Add forwarding info
    client_ip = request.headers.get("X-Forwarded-For")
    if not client_ip and request.client:
        client_ip = request.client.host
    if client_ip:
        headers["X-Forwarded-For"] = client_ip
    
    headers["X-Forwarded-Proto"] = request.url.scheme
    headers["X-Forwarded-Host"] = request.headers.get("Host", request.url.netloc)
    
    return headers


def get_target_service(path: str) -> Optional[str]:
    """
    Determine the target service based on request path.
    
    Args:
        path: The request path.
    
    Returns:
        Optional[str]: The target service name, or None if no match.
    """
    for prefix, service in SERVICE_ROUTES.items():
        if path.startswith(prefix):
            return service
    return None


def transform_path(path: str, service: str) -> str:
    """
    Transform the request path for the backend service.
    
    By default, removes the service prefix from the path.
    
    Args:
        path: The original request path.
        service: The target service name.
    
    Returns:
        str: The transformed path.
    """
    # Find the matching prefix
    for prefix, svc in SERVICE_ROUTES.items():
        if svc == service and path.startswith(prefix):
            # Remove the prefix
            transformed = path[len(prefix):] or "/"
            return transformed
    
    return path


async def proxy_request(
    request: Request,
    service: str,
    target_path: str,
) -> Response:
    """
    Proxy a request to a backend service.
    
    Args:
        request: The incoming request.
        service: Target service name.
        target_path: Path on the target service.
    
    Returns:
        Response: The proxied response.
    """
    app_state = getattr(request.app.state, "app_state", None)
    http_client = getattr(app_state, "http_client", None)
    service_discovery: ServiceDiscovery = getattr(app_state, "service_discovery", None)
    
    if not http_client:
        raise ServiceUnavailable(
            service="gateway",
            message="HTTP client not initialized",
        )
    
    # Get service endpoint
    if service_discovery:
        endpoint = await service_discovery.get_endpoint(service)
    else:
        # Fallback to static configuration
        settings = getattr(app_state, "settings", None)
        endpoint = getattr(settings.backend, f"{service.replace('-', '_')}_url", None)
    
    if not endpoint:
        raise NotFoundError(
            resource_type="service",
            resource_id=service,
            message=f"Service '{service}' not found",
        )
    
    # Build target URL
    target_url = f"{endpoint}{target_path}"
    if request.url.query:
        target_url = f"{target_url}?{request.url.query}"
    
    # Build headers
    headers = get_proxy_headers(request)
    
    # Get request body
    body = await request.body()
    
    # Log the proxy request
    logger.debug(
        "Proxying request",
        service=service,
        method=request.method,
        target_url=target_url,
        request_id=getattr(request.state, "request_id", None),
    )
    
    start_time = time.perf_counter()
    
    try:
        # Make the request with circuit breaker protection
        async def make_request():
            return await http_client.request(
                method=request.method,
                url=target_url,
                headers=headers,
                content=body if body else None,
                follow_redirects=False,
            )
        
        response = await with_circuit_breaker(
            service,
            make_request,
        )
        
        duration = time.perf_counter() - start_time
        
        # Record metrics
        record_upstream_duration(
            service=service,
            method=request.method,
            status=response.status_code,
            duration=duration,
        )
        
        logger.debug(
            "Upstream response received",
            service=service,
            status_code=response.status_code,
            duration_ms=round(duration * 1000, 2),
            request_id=getattr(request.state, "request_id", None),
        )
        
        # Build response headers (exclude hop-by-hop headers)
        response_headers = {}
        for key, value in response.headers.items():
            key_lower = key.lower()
            if key_lower not in EXCLUDED_HEADERS:
                response_headers[key] = value
        
        # Add gateway headers
        response_headers["X-Gateway-Request-Id"] = getattr(
            request.state, "request_id", "unknown"
        )
        response_headers["X-Upstream-Service"] = service
        response_headers["X-Upstream-Duration-Ms"] = str(round(duration * 1000, 2))
        
        # Return streaming response for large bodies
        if response.headers.get("content-length"):
            content_length = int(response.headers.get("content-length", 0))
            if content_length > 1024 * 1024:  # 1MB threshold
                return StreamingResponse(
                    content=response.aiter_bytes(),
                    status_code=response.status_code,
                    headers=response_headers,
                    media_type=response.headers.get("content-type"),
                )
        
        return Response(
            content=response.content,
            status_code=response.status_code,
            headers=response_headers,
            media_type=response.headers.get("content-type"),
        )
        
    except httpx.TimeoutException as e:
        duration = time.perf_counter() - start_time
        record_upstream_duration(
            service=service,
            method=request.method,
            status=504,
            duration=duration,
        )
        
        logger.error(
            "Upstream request timeout",
            service=service,
            target_url=target_url,
            duration_ms=round(duration * 1000, 2),
            request_id=getattr(request.state, "request_id", None),
        )
        
        raise TimeoutError(
            service=service,
            timeout=duration,
        )
        
    except httpx.ConnectError as e:
        logger.error(
            "Failed to connect to upstream service",
            service=service,
            target_url=target_url,
            error=str(e),
            request_id=getattr(request.state, "request_id", None),
        )
        
        raise ServiceUnavailable(
            service=service,
            message=f"Failed to connect: {str(e)}",
        )
        
    except Exception as e:
        logger.exception(
            "Unexpected error proxying request",
            service=service,
            target_url=target_url,
            request_id=getattr(request.state, "request_id", None),
        )
        raise


@router.api_route(
    "/{path:path}",
    methods=["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"],
    include_in_schema=False,
)
async def proxy_all(request: Request, path: str) -> Response:
    """
    Catch-all proxy endpoint.
    
    Routes all requests that don't match specific endpoints
    to the appropriate backend service.
    
    Args:
        request: The incoming request.
        path: The request path.
    
    Returns:
        Response: The proxied response.
    """
    # Add leading slash
    full_path = f"/{path}"
    
    # Find target service
    service = get_target_service(full_path)
    
    if not service:
        raise NotFoundError(
            message=f"No service found for path: {full_path}",
            details={"path": full_path},
        )
    
    # Transform path for backend
    target_path = transform_path(full_path, service)
    
    return await proxy_request(request, service, target_path)

File: nexus-platform/services/api-gateway/go/internal/proxy/proxy.go
// =============================================================================
// NEXUS PLATFORM - API GATEWAY - REVERSE PROXY
// =============================================================================
// Reverse proxy implementation for routing requests to backend services.
// =============================================================================

package proxy

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/rs/zerolog"

	"github.com/nexus-platform/api-gateway/internal/config"
	"github.com/nexus-platform/api-gateway/internal/middleware"
	"github.com/nexus-platform/api-gateway/internal/service"
)

// Config holds proxy configuration.
type Config struct {
	HTTPClient      *http.Client
	ServiceRegistry *service.Registry
	Logger          zerolog.Logger
	Config          *config.Config
}

// Handler is the reverse proxy handler.
type Handler struct {
	httpClient      *http.Client
	serviceRegistry *service.Registry
	logger          zerolog.Logger
	config          *config.Config
}

// ServiceRoutes maps URL prefixes to backend services.
var ServiceRoutes = map[string]string{
	"/auth":      "auth-service",
	"/users":     "user-service",
	"/payments":  "payment-service",
	"/workflows": "workflow-engine",
	"/queue":     "distributed-queue",
	"/stream":    "streaming-platform",
	"/cache":     "cache-layer",
	"/scheduler": "scheduler-service",
	"/workers":   "worker-fleet",
}

// ForwardHeaders are headers to forward from client to backend.
var ForwardHeaders = map[string]bool{
	"accept":           true,
	"accept-encoding":  true,
	"accept-language":  true,
	"content-type":     true,
	"content-length":   true,
	"user-agent":       true,
	"x-request-id":     true,
	"x-correlation-id": true,
	"x-forwarded-for":  true,
	"x-forwarded-proto": true,
	"x-real-ip":        true,
	"authorization":    true,
}

// ExcludedHeaders are hop-by-hop headers to exclude.
var ExcludedHeaders = map[string]bool{
	"connection":       true,
	"keep-alive":       true,
	"transfer-encoding": true,
	"te":               true,
	"trailer":          true,
	"upgrade":          true,
	"host":             true,
}

// NewHandler creates a new proxy handler.
func NewHandler(cfg Config) *Handler {
	return &Handler{
		httpClient:      cfg.HTTPClient,
		serviceRegistry: cfg.ServiceRegistry,
		logger:          cfg.Logger,
		config:          cfg.Config,
	}
}

// ServeHTTP handles proxy requests.
func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	requestID := middleware.GetRequestID(ctx)

	// Determine target service
	serviceName := h.getTargetService(r.URL.Path)
	if serviceName == "" {
		h.writeError(w, http.StatusNotFound, "NOT_FOUND", "No service found for path")
		return
	}

	// Get service endpoint
	endpoint := h.serviceRegistry.GetEndpoint(serviceName)
	if endpoint == "" {
		h.logger.Error().
			Str("request_id", requestID).
			Str("service", serviceName).
			Msg("Service endpoint not found")
		h.writeError(w, http.StatusServiceUnavailable, "SERVICE_UNAVAILABLE", fmt.Sprintf("Service '%s' not available", serviceName))
		return
	}

	// Transform path
	targetPath := h.transformPath(r.URL.Path, serviceName)

	// Build target URL
	targetURL := endpoint + targetPath
	if r.URL.RawQuery != "" {
		targetURL = targetURL + "?" + r.URL.RawQuery
	}

	// Create proxy request
	proxyReq, err := http.NewRequestWithContext(ctx, r.Method, targetURL, r.Body)
	if err != nil {
		h.logger.Error().
			Err(err).
			Str("request_id", requestID).
			Msg("Failed to create proxy request")
		h.writeError(w, http.StatusInternalServerError, "INTERNAL_ERROR", "Failed to create proxy request")
		return
	}

	// Copy headers
	h.copyHeaders(proxyReq, r)

	// Add gateway headers
	h.addGatewayHeaders(proxyReq, r)

	// Log proxy request
	h.logger.Debug().
		Str("request_id", requestID).
		Str("service", serviceName).
		Str("method", r.Method).
		Str("target_url", targetURL).
		Msg("Proxying request")

	startTime := time.Now()

	// Execute request
	resp, err := h.httpClient.Do(proxyReq)
	if err != nil {
		duration := time.Since(startTime)
		h.logger.Error().
			Err(err).
			Str("request_id", requestID).
			Str("service", serviceName).
			Dur("duration", duration).
			Msg("Upstream request failed")
		h.writeError(w, http.StatusBadGateway, "BAD_GATEWAY", "Failed to connect to upstream service")
		return
	}
	defer resp.Body.Close()

	duration := time.Since(startTime)

	h.logger.Debug().
		Str("request_id", requestID).
		Str("service", serviceName).
		Int("status", resp.StatusCode).
		Dur("duration", duration).
		Msg("Upstream response received")

	// Copy response headers
	h.copyResponseHeaders(w, resp)

	// Add gateway response headers
	w.Header().Set("X-Gateway-Request-Id", requestID)
	w.Header().Set("X-Upstream-Service", serviceName)
	w.Header().Set("X-Upstream-Duration-Ms", fmt.Sprintf("%.2f", float64(duration.Microseconds())/1000))

	// Write status code
	w.WriteHeader(resp.StatusCode)

	// Copy response body
	io.Copy(w, resp.Body)
}

// getTargetService determines the target service based on path.
func (h *Handler) getTargetService(path string) string {
	for prefix, service := range ServiceRoutes {
		if strings.HasPrefix(path, prefix) {
			return service
		}
	}
	return ""
}

// transformPath transforms the request path for the backend.
func (h *Handler) transformPath(path string, serviceName string) string {
	for prefix, svc := range ServiceRoutes {
		if svc == serviceName && strings.HasPrefix(path, prefix) {
			// Remove the prefix
			transformed := strings.TrimPrefix(path, prefix)
			if transformed == "" {
				transformed = "/"
			}
			return transformed
		}
	}
	return path
}

// copyHeaders copies headers from the original request to the proxy request.
func (h *Handler) copyHeaders(proxyReq *http.Request, r *http.Request) {
	for key, values := range r.Header {
		keyLower := strings.ToLower(key)
		if ExcludedHeaders[keyLower] {
			continue
		}
		if ForwardHeaders[keyLower] {
			for _, value := range values {
				proxyReq.Header.Add(key, value)
			}
		}
	}
}

// addGatewayHeaders adds gateway-specific headers to the request.
func (h *Handler) addGatewayHeaders(proxyReq *http.Request, r *http.Request) {
	ctx := r.Context()
	requestID := middleware.GetRequestID(ctx)

	proxyReq.Header.Set("X-Gateway-Request-Id", requestID)

	// Add user info if authenticated
	if user := middleware.GetUser(ctx); user != nil {
		proxyReq.Header.Set("X-User-Id", user.ID)
		if user.TenantID != "" {
			proxyReq.Header.Set("X-Tenant-Id", user.TenantID)
		}
		if len(user.Roles) > 0 {
			proxyReq.Header.Set("X-User-Roles", strings.Join(user.Roles, ","))
		}
	}

	// Add forwarding headers
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		proxyReq.Header.Set("X-Forwarded-For", xff)
	} else if r.RemoteAddr != "" {
		proxyReq.Header.Set("X-Forwarded-For", strings.Split(r.RemoteAddr, ":")[0])
	}

	proxyReq.Header.Set("X-Forwarded-Proto", h.getScheme(r))
	proxyReq.Header.Set("X-Forwarded-Host", r.Host)
}

// getScheme returns the request scheme.
func (h *Handler) getScheme(r *http.Request) string {
	if r.TLS != nil {
		return "https"
	}
	if proto := r.Header.Get("X-Forwarded-Proto"); proto != "" {
		return proto
	}
	return "http"
}

// copyResponseHeaders copies headers from the upstream response.
func (h *Handler) copyResponseHeaders(w http.ResponseWriter, resp *http.Response) {
	for key, values := range resp.Header {
		keyLower := strings.ToLower(key)
		if ExcludedHeaders[keyLower] {
			continue
		}
		for _, value := range values {
			w.Header().Add(key, value)
		}
	}
}

// writeError writes an error response.
func (h *Handler) writeError(w http.ResponseWriter, status int, code, message string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	fmt.Fprintf(w, `{"error":{"code":"%s","message":"%s"}}`, code, message)
}
